src/processing/spot.rs

```diff
@@ -217,53 +218,83 @@ pub async fn process_long_tail_asset(
     pair: String,
     sources: Vec<String>,
 ) -> Result<u64, MonitoringError> {
+    let config = get_config(None).await;
+    let network_env = &config.network_str();
+    let decimals = *config.decimals(DataType::Spot).get(&pair).unwrap();
+
+    let mut timestamps = Vec::with_capacity(sources.len());
+    let mut deviations = Vec::with_capacity(sources.len());
+
+    for source in sources.iter() {
+        log::info!("Processing data for pair: {} and source: {}", pair, source);
+        let (timestamp, price_deviation) =
+            get_price_deviation_for_source_from_chain(pool.clone(), &pair, source, decimals)
+                .await?;
+        timestamps.push(timestamp);
+        deviations.push(price_deviation);
+    }
+
+    let threshold = get_long_tail_threshold(&pair, sources.len()).unwrap();
+
+    // TODO: Maybe we should only consider recent sources, i.e in the last hour?
+    // Count deviating sources
+    let deviating_sources = deviations
+        .iter()
+        .filter(|&&deviation| deviation >= threshold)
+        .count();
+
+    // Set the metric for the number of deviating sources
+    LONG_TAIL_ASSET_DEVIATING_SOURCES
+        .with_label_values(&[network_env, &pair, "spot"])
+        .set(deviating_sources as f64);
+
+    // Set the metric for the total number of sources
+    LONG_TAIL_ASSET_TOTAL_SOURCES
+        .with_label_values(&[network_env, &pair, "spot"])
+        .set(sources.len() as f64);
+
+    Ok(timestamps.last().copied().unwrap())
+}
+
+pub async fn get_price_deviation_for_source_from_chain(
+    pool: deadpool::managed::Pool<AsyncDieselConnectionManager<AsyncPgConnection>>,
+    pair: &str,
+    source: &str,
+    decimals: u32,
+) -> Result<(u64, f64), MonitoringError> {
```

Comment:
let's return a struct here 
```suggestion
) -> Result<Deviation, MonitoringError> {
```

Reply:
done